src/cljc/rems/common/form.cljc

```diff
@@ -62,3 +79,297 @@
+  (letfn [(validate-field [index field]
+            {index (merge (validate-field-type field)
+                          (validate-localized-text-field field :field/title languages)
+                          (when (supports-placeholder? field)
+                            (validate-optional-localized-field field :field/placeholder languages))
+                          (when (supports-max-length? field)
+                            (validate-max-length (:field/max-length field)))
+                          (when (supports-options? field)
+                            (validate-options (:field/options field) languages))
+                          (when (supports-visibility? field)
+                            (validate-visibility field fields)))})]
+    (apply merge (map-indexed validate-field fields))))
+
+(defn- nil-if-empty [m]
+  (when-not (empty? m)
+    m))
+
+(defn validate-form [form languages]
+  (-> (merge (validate-text-field form :form/organization)
+             (validate-text-field form :form/title)
+             {:form/fields (validate-fields (:form/fields form) languages)})
+      remove-empty-keys
+      nil-if-empty))
+
+(deftest validate-form-test
+  (let [form {:form/organization "abc"
+              :form/title "the title"
+              :form/fields [{:field/id "fld1"
+                             :field/title {:en "en title"
+                                           :fi "fi title"}
+                             :field/optional true
+                             :field/type :text
+                             :field/max-length "12"
+                             :field/placeholder {:en "en placeholder"
+                                                 :fi "fi placeholder"}}]}
+        languages [:en :fi]]
+
+    (testing "valid form"
+      (is (empty? (validate-form form languages))))
+
+    (testing "missing organization"
+      (is (= (:form/organization (validate-form (assoc-in form [:form/organization] "") languages))
+             :t.form.validation/required)))
+
+    (testing "missing title"
+      (is (= (:form/title (validate-form (assoc-in form [:form/title] "") languages))
+             :t.form.validation/required)))
+
+    (testing "zero fields is ok"
+      (is (empty? (validate-form (assoc-in form [:form/fields] []) languages))))
+
+    (testing "missing field title"
+      (let [nil-title (validate-form (assoc-in form [:form/fields 0 :field/title] nil) languages)]
+        (is (= (get-in (validate-form (assoc-in form [:form/fields 0 :field/title :en] "") languages)
+                       [:form/fields 0 :field/title :en])
+               (get-in (validate-form (update-in form [:form/fields 0 :field/title] dissoc :en) languages)
+                       [:form/fields 0 :field/title :en])
+               (get-in nil-title [:form/fields 0 :field/title :en])
+               (get-in nil-title [:form/fields 0 :field/title :fi])
+               :t.form.validation/required))))
```

Comment:
could you remove the get-ins here (and other places) and assert the whole return value of validate-form? then the test would act better as documentation IMO

Reply:
I inherited this part so I intended not to touch it yet. I'll add a cleanup issue about it.